### PR TITLE
clientv3: use DialContext

### LINKED ISSUE #7216: client called Close() but some goroutine not exit
Hi, 

I use a simple test case with clientv3 API like below, and find that one goroutine leak until 20s after client.Close():
1. use docker container to start a 3 nodes etcd cluster
2. new a clientv3 with all nodes' endpoints
3. one goroutine put sub key 
4. one goroutine crash one etcd server(docker rm the etcd container)
5. then terminate all the etcd server and close client

leak stack like this:
```
Too many goroutines running after all test(s).
1 instances of:
net.(*netFD).connect.func2(...)
        /usr/local/go/src/net/fd_unix.go:129 +0x1d5
created by net.(*netFD).connect
        /usr/local/go/src/net/fd_unix.go:140 +0x255
1 instances of:
net.runtime_pollWait(...)
        /usr/local/go/src/runtime/netpoll.go:160 +0x59
net.(*pollDesc).wait(...)
        /usr/local/go/src/net/fd_poll_runtime.go:73 +0x38
net.(*pollDesc).waitWrite(...)
        /usr/local/go/src/net/fd_poll_runtime.go:82 +0x34
net.(*netFD).connect(...)
        /usr/local/go/src/net/fd_unix.go:152 +0x26a
net.(*netFD).dial(...)
        /usr/local/go/src/net/sock_posix.go:137 +0x137
net.socket(0xbd3860, 0xc4202d1c20, 0x95b6e9, 0x3, 0x2, 0x1, 0x0, 0x0, 0xbd4640, 0x0, ...)
        /usr/local/go/src/net/sock_posix.go:89 +0x209
net.internetSocket(0xbd3860, 0xc4202d1c20, 0x95b6e9, 0x3, 0xbd4640, 0x0, 0xbd4640, 0xc4202d89f0, 0x1, 0x0, ...)
        /usr/local/go/src/net/ipsock_posix.go:158 +0x129
net.doDialTCP(...)
        /usr/local/go/src/net/tcpsock_posix.go:58 +0xb9
net.dialTCP(...)
        /usr/local/go/src/net/tcpsock_posix.go:54 +0xe4
net.dialSingle(...)
        /usr/local/go/src/net/dial.go:501 +0x479
net.dialSerial(...)
        /usr/local/go/src/net/dial.go:469 +0x223
net.(*Dialer).DialContext(0xc4204575c0, 0xbd3820, 0xc420016620, 0x95b6e9, 0x3, 0xc420144a60, 0x10, 0x0, 0x0, 0x0, ...)
        /usr/local/go/src/net/dial.go:351 +0x657
net.(*Dialer).Dial(...)
        /usr/local/go/src/net/dial.go:282 +0x75
net.DialTimeout(...)
        /usr/local/go/src/net/dial.go:268 +0x9d
/xxxx/vendor/github.com/coreos/etcd/clientv3.(*Client).dialSetupOpts.func1(...)
        /xxxx/vendor/github.com/coreos/etcd/clientv3/client.go:169 +0x1ae
/xxxx/vendor/google.golang.org/grpc.WithDialer.func1.1(...)
        /xxxx/vendor/google.golang.org/grpc/clientconn.go:204 +0xd1
/xxxx/vendor/google.golang.org/grpc/transport.dial(...)
        /xxxx/vendor/google.golang.org/grpc/transport/http2_client.go:112 +0x5a
/xxxx/vendor/google.golang.org/grpc/transport.newHTTP2Client(0x7ff2d7c6a258, 0xc4202d1b60, 0xc420144a60, 0x10, 0x0, 0x0, 0xc420114bc0, 0x0, 0x0, 0x0, ...)
        /xxxx/vendor/google.golang.org/grpc/transport/http2_client.go:122 +0xbf
/xxxx/vendor/google.golang.org/grpc/transport.NewClientTransport(0x7ff2d7c6a258, 0xc4202d1b60, 0xc420144a60, 0x10, 0x0, 0x0, 0xc420114bc0, 0x0, 0x0, 0x0, ...)
        /xxxx/vendor/google.golang.org/grpc/transport/transport.go:368 +0x87
/xxxx/vendor/google.golang.org/grpc.(*addrConn).resetTransport(...)
        /xxxx/vendor/google.golang.org/grpc/clientconn.go:648 +0x25e
/xxxx/vendor/google.golang.org/grpc.(*addrConn).transportMonitor(...)
        /xxxx/vendor/google.golang.org/grpc/clientconn.go:751 +0x362
/xxxx/vendor/google.golang.org/grpc.(*ClientConn).resetAddrConn.func1(...)
        /xxxx/vendor/google.golang.org/grpc/clientconn.go:467 +0x1de
created by /xxxx/vendor/google.golang.org/grpc.(*ClientConn).resetAddrConn
        /xxxx/vendor/google.golang.org/grpc/clientconn.go:468 +0x3c7
ERROR: *utils.EtcdWatcherTestSuite is failed by leak goroutine check.
ERROR: *utils.EtcdWatcherTestSuite is failed by leak goroutine check.
```

case code like this:
```
const (
	NodeCount = 3
	ClientDialTimeout = 1
        SleepTimeInSec = 20
)

type EtcdWatcherTestSuite struct {
	suite.LambdaSuite  // LambdaSuite can CheckLeakedGoroutine
	cluster *etcd.Cluster
	// the client has all node endpoint
	client *clientv3.Client
}

func (s *EtcdWatcherTestSuite) Init() {
	assert := s.Require()
	//flag.Lookup("stderrthreshold").Value.Set("INFO")
	var err error
	s.cluster, err = etcd.NewCluster(NodeCount)
	assert.Nil(err)
	endpoints := []string{}
	for i := 0; i < NodeCount; i++ {
		eps := s.cluster.Client(i).Endpoints()
		endpoints = append(endpoints, eps...)
	}
	logger.Infof("Total endpoints %v\n", endpoints)
	s.client, err = clientv3.New(clientv3.Config{
		Endpoints: endpoints,
		DialTimeout: ClientDialTimeout * time.Second})
	assert.Nil(err)
}

func (s *EtcdWatcherTestSuite) Finish() {
	s.client.Close()
	s.cluster.Terminate()
	time.Sleep(SleepTimeInSec*time.Second)
	logger.Flush()
}

func (s *EtcdWatcherTestSuite) TestWatchWithCrash() {
	s.Init()
	defer s.Finish()

	stop := make(chan int)
	var wg sync.WaitGroup
	wg.Add(2)
	go func() {
		defer wg.Done()
		max := big.NewInt(int64(NodeCount))
		i, _ := rand.Int(rand.Reader, max)
		crashClient := s.cluster.CrashMember(int(i.Uint64()))
		defer crashClient.Close()
		logger.Infof("Crash node %d\n", i.Uint64())
	}()

	go func() {
		count := 0
		defer wg.Done()
		for {
			select {
			case <-time.After(10 * time.Millisecond):
				ctx, cancel := context.WithTimeout(context.Background(), time.Second)
				countStr := strconv.Itoa(count)
				s.client.Put(ctx, "/EtcdWatcherTest/TestWatchWithFailover/testKey_" + countStr, countStr)
				cancel()
				count = count + 1

			case <-stop:
				logger.Infof("Put loop exit\n")
				return
			}
		}
	}()

	go func() {
		select {
		case <-time.After(5 * time.Second):
			close(stop)
		}
	}()

	wg.Wait()
}

func TestEtcdWatcher(t *testing.T) {
	suite.Run(t, new(EtcdWatcherTestSuite))
}
```
etcd node start like this : 

```
"docker run -d --net=%s --ip=%s --name %s " + EtcdImagePath +
			" /usr/local/bin/etcd " +
			"--data-dir=data.etcd --name %s " +
			"--initial-advertise-peer-urls http://%s:2380 " +
			"--listen-peer-urls http://%s:2380 " +
			"--advertise-client-urls http://%s:2379 " +
			"--listen-client-urls http://%s:2379 " +
			"--initial-cluster " + cluster + " " +
			"--initial-cluster-state " + clusterState + " " +
			"--initial-cluster-token " + token
"
```
crash like this:
```
// CrashMember : rm one member of the multi-mode etcd cluster
func (c *Cluster) CrashMember(i int) *clientv3.Client{
	cmd := fmt.Sprintf("docker rm -f %s", c.members[i].name)
	out, err := exec.Command("sh", "-c", cmd).CombinedOutput()
	if err != nil {
		glog.Errorf("Failed to remove container %s. Message: %s. Error: %v.", c.members[i].name, out, err)
	} else {
		glog.Infof("Succeed to remove container %s.", c.members[i].name)
	}
	retClient := c.clients[i]
	c.members = append(c.members[:i], c.members[i+1:]...)
	c.clients = append(c.clients[:i], c.clients[i+1:]...)
	return retClient
}
```
terminate like this:
```
// Terminate : terminate cluster
func (c *Cluster) Terminate() {
	var cmd string

	// Close the clients.
	for i := 0; i < len(c.clients); i++ {
		c.clients[i].Close()
	}

	// Cleanup the containers.
	for i := 0; i < len(c.members); i++ {
		name := c.members[i].name
		cmd = fmt.Sprintf("docker rm -f %s", name)
		out, err := exec.Command("sh", "-c", cmd).CombinedOutput()
		if err != nil {
			glog.Errorf("Failed to remove container %s. Error: %v. Message: %s.", name, err, out)
		} else {
			glog.Infof("Succeed to remove container %s", name)
		}
	}
```

etcd vesion: 3.0.16-a23109a0c6797119034e613af435ecfaa09e5723<br>

request:
1. why it can't  cancel every operation of client immediately when Close() is called?
2. if DialTimeout set to be 1second, it needs to sleep 20s; and DialTimeout set to be 5 seconds, it needs to sleep 60s? what the relativeship between DialTimeout and the goroutine exit time?

### PR DESCRIPTION
**Pull request by heyitsanthony**
Fixes #7216

### CODE REVIEW

**Comment by xiang90** [comment]
Can we add a test for this?

File: clientv3/client_test.go [hunk 1]
```diff
@@ -25,6 +26,47 @@ import (
...
+
+	// issue Get to force redial attempts
+	go c.Get(context.TODO(), "abc")
+
+	// wait a little bit so client close is after dial starts
+	time.Sleep(100 * time.Millisecond)
+
+	donec := make(chan struct{})
+	go func() {
+		close(donec)
```
Comment (xiang90) [inline clientv3/client_test.go]: we should close(donec) after we call c.Close() ?

**Comment by xiang90** [comment]
lgtm

File: clientv3/client_test.go (hunk 1 above)
Comment (heyitsanthony) [inline clientv3/client_test.go]: oh oops, was getting the leak even without this...

**Pull request merged**